### PR TITLE
Renaming Filters

### LINKED ISSUE #5207: Allow custom naming in tables / graph series
## Is your feature request related to a problem?

In trends, when I create a graph / table, PostHog tries to be clever and come up with names for me.

I.e 'Event name (Active feature flags = some-feature)'

This can be confusing, as there are often some weirdities in these names I.e the symbols for contains / does not contain are so similar, and when in the dashboard, users ask me what they mean.

## Describe the solution you'd like

By default, use the name PostHog suggests, but allow the user to rename the series as they require.

## Describe alternatives you've considered

Don't allow renaming and stick with the name PostHog suggests?

## Additional context

n/a

#### *Thank you* for your feature request – we love each and every one!

### LINKED ISSUE #5229: Better tooltips for filtered trends
## Is your feature request related to a problem?

Dashboard trends are impossible to understand when they have a breakdown applied.

Screenshot 2021-07-20 at 17 23 04

This is from an [internal marketing dashboard](https://app.posthog.com/dashboard/7040) - with the [full insight trend here](https://app.posthog.com/insights?insight=TRENDS&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%2C%22properties%22%3A%5B%7B%22key%22%3A%22%24current_url%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22posthog.com%2Fblog%2F%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22%24pageview%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A1%2C%22properties%22%3A%5B%7B%22key%22%3A%22%24current_url%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%5B%22https%3A%2F%2Fposthog.com%2F%22%5D%2C%22operator%22%3A%22exact%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22%24pageview%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A2%2C%22properties%22%3A%5B%7B%22key%22%3A%22%24current_url%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22posthog.com%2Fcareers%2F%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22%24pageview%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A3%2C%22properties%22%3A%5B%7B%22key%22%3A%22%24current_url%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22posthog.com%2Fhandbook%2F%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22%24pageview%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A4%2C%22properties%22%3A%5B%7B%22key%22%3A%22%24current_url%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22posthog.com%2Fdocs%2F%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22math_property%22%3Anull%7D%5D&display=ActionsLineGraph&interval=week&date_from=yStart&properties=%5B%7B%22key%22%3A%22%24current_url%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22https%3A%2F%2Fapp.posthog.com%2F%22%2C%22operator%22%3A%22not_icontains%22%7D%5D&filter_test_accounts=true#fromItem=271130&fromItemName=Site%20Visitors%20-%20Weekly%20Uniques&fromDashboard=7040). 

I can't tell what the meaning of the different lines is.
 
## Describe the solution you'd like

The tooltip to make clearer what the different lines are - when there are multiple graph series that are differentiated based on the filter applied, then add text to explain the filter to the tooltip, _or_ a legend to appear when there are multiple lines on a trend in the dashboard.

## Describe alternatives you've considered

Click into the dashboard item to look at the chart. The tooltips aren't perfect for this - they still don't expand at all, BUT you can use them to figure out which line is which by looking at the letters in front of the graph series:

Screenshot 2021-07-20 at 17 27 21

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

Adds the ability to rename entity filters for all of our insights. This PR implements the frontend side of this. All new features are hidden behind a FF called `6063-rename-filters`.

Backend PR: #6090

Closes #5229.
Closes #5207.

~~Note: This PR will be rebased on `master` once #6090 is in. You can ignore diffs in all *.py files~~

## Screenshots

**Demo**

https://user-images.githubusercontent.com/13460330/134576038-ead070e8-2a56-48ca-88a5-bdcc0521ea04.mov

A few edge cases below. Note, I probably haven't tested all edge cases so make sure to click around and try to break things.

**Trends + Legend**

Screen Shot 2021-09-23 at 12 41 09 PM

**Trends table**

Screen Shot 2021-09-23 at 12 41 13 PM

**Dashboards**

Screen Shot 2021-09-23 at 12 57 43 PM

**Saved Insight**

Screen Shot 2021-09-23 at 12 57 13 PM

**With vertical breakdowns**

Screen Shot 2021-09-23 at 12 40 54 PM

**Without vertical breakdowns**

Screen Shot 2021-09-23 at 12 44 02 PM

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [x] Jest frontend tests
- [x] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [x] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Review by clarkus — COMMENTED** [review]
I tried using the deployment to review these changes, but doesn't quite seem to be working. The modal never triggers to facilitate the renaming.

**Comment by alexkim205** [comment]
Ah sorry Chris, this is still in draft mode so I haven't implemented the modal yet. I should've mentioned this in the CR description. I'll bump you again once it's ready for review, sorry for the inconvenience!

**Comment by clarkus** [comment]
No worries. Thanks!

**Comment by alexkim205** [comment]
Ready! retagging for reviews @clarkus @mariusandra @Twixes 

@mariusandra My first stab at logic tests so please lmk if it can be improved!

**Review by clarkus — DISMISSED** [review]
Looks good to me! The one change I think we should consider before merging is shown in this screenshot. When you trigger the rename, you're presented with the modal and the name field. There's a default value there. I think if we focused and selected the field value, it'd be easy for a user to use key commands to replace, amend, whatever the text there. It would just be one less action needed in order to rename the step. 

Screen Shot 2021-09-23 at 3 17 03 PM

**Comment by alexkim205** [comment]
Autofocusing now works

https://user-images.githubusercontent.com/13460330/134596033-14ee2756-316e-46cf-a11b-583767500df1.mov

**Review by clarkus — APPROVED** [review]

File: frontend/src/scenes/insights/utils.ts [hunk 1]
```diff
@@ -19,3 +19,12 @@ export const getLogicFromInsight = (
         return trendsLogic(logicProps)
     }
 }
+
+export const getDisplayNameFromEntityFilter = (
+    filter: EntityFilter | ActionFilter | null,
+    isCustom = true
+): string | null => {
+    const customName = (filter?.custom_name ?? '').trim() === '' ? null : filter?.custom_name
```
Comment (rcmarron) [inline frontend/src/scenes/insights/utils.ts:27]: Nit: this took me a bit to understand what was happening. Might be good to break it out into a few lines.

**Review by rcmarron — DISMISSED** [review]
Overall, looks great! Awesome feature!

A couple of bugs/thoughts:
1) On vertical funnels, if you rename later steps, the custom names don't appear in the graph below:
image

2) If I rename the first step, and then I add additional steps, the new steps automatically get the custom name of the first step. I'd expect the new steps to have the default name (e.g. $pageview)

3) It doesn't seem like we use the word "query step" elsewhere in the product. I wonder if we should keep it consistent with "graph series" and "funnel step" on the respective page.
image

4) On funnels, it might make sense to include the custom names in the "time to convert" and "Historical trends" drop downs.
image

5) I had some funny behavior around custom names not appearing after saving an insight, but I can't repro' it so.... maybe just a fluke.

File: frontend/src/lib/components/EntityFilterInfo.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+export function EntityFilterInfo({ filter }: Props): JSX.Element {
+    const title = getDisplayNameFromEntityFilter(filter)
+    const subtitle = getDisplayNameFromEntityFilter(filter, false)
+
+    if (filter.type === EntityTypes.NEW_ENTITY || (!title && !subtitle)) {
+        return <TextWrapper title="Select filter">Select filter</TextWrapper>
+    }
+
+    const titleTranslation = getKeyMapping(title, 'event')?.label ?? title ?? undefined
+    const subTitleTranslation = getKeyMapping(subtitle, 'event')?.label ?? subtitle ?? undefined
```
Comment (rcmarron) [inline frontend/src/lib/components/EntityFilterInfo.tsx:29]: Translation feels like a misnomer here (particularly with strings). Maybe `titleToDisplay`?

**Comment by alexkim205** [comment]
Thanks for the comprehensive review Rick!

> 1. On vertical funnels, if you rename later steps, the custom names don't appear in the graph below:

> 5. I had some funny behavior around custom names not appearing after saving an insight, but I can't repro' it so.... maybe just a fluke.

I think this might be happening bc these filter names are cached on the backend (which could also be related to point 5). Could you try hard refreshing the graph a bunch of times and see if this fixes things?

> 2. If I rename the first step, and then I add additional steps, the new steps automatically get the custom name of the first step. I'd expect the new steps to have the default name (e.g. $pageview)

Good catch, fixed!

> 3. It doesn't seem like we use the word "query step" elsewhere in the product. I wonder if we should keep it consistent with "graph series" and "funnel step" on the respective page.

Changed!

> 4. On funnels, it might make sense to include the custom names in the "time to convert" and "Historical trends" drop downs.

Good call. I think it'd be good to include the custom names in the dropdown menu. I'll exclude them in the input field itself to avoid making that label variably longer.

File: frontend/src/lib/components/EntityFilterInfo.tsx (hunk 2 above)
Comment (alexkim205) [reply frontend/src/lib/components/EntityFilterInfo.tsx:29]: sounds much better. english is hard

**Comment by alexkim205** [comment]
As this already went through one round of code reviews, I'm going to merge this one in for now so that it's not out for too long. This is hidden behind a FF so it should be a pretty safe deployment. I'll address any followup bugs in future PR's so feel free to keep the convo going here.

**Comment by joesaunderson** [comment]
Love this!!

File: frontend/src/scenes/insights/ActionFilter/ActionFilter.tsx [hunk 3]
```diff
@@ -94,6 +97,8 @@ export const ActionFilter = React.forwardRef<HTMLDivElement, ActionFilterProps>(
         const { localFilters } = useValues(logic)
         const { addFilter, setLocalFilters } = useActions(logic)
 
+        const [modalOpen, setModalOpen] = useState(false)
+
         // No way around this. Somehow the ordering of the logic calling each other causes stale "localFilters"
```
Comment (mariusandra) [inline frontend/src/scenes/insights/ActionFilter/ActionFilter.tsx:102]: Can't this state be moved into the entity filter logic?

File: frontend/src/scenes/insights/ActionFilter/FilterRenameModal.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+export function FilterRenameModal({ visible, setModalOpen, view }: Props): JSX.Element {
+    const ref = useRef<Input | null>(null)
+    const { selectedFilter } = useValues(entityFilterLogic)
+    const { renameFilter, selectFilter } = useActions(entityFilterLogic)
+
+    const [name, setName] = useState(getDisplayNameFromEntityFilter(selectedFilter) ?? '')
+
+    useEffect(() => {
+        setName(getDisplayNameFromEntityFilter(selectedFilter) ?? '')
+    }, [selectedFilter])
```
Comment (mariusandra) [inline frontend/src/scenes/insights/ActionFilter/FilterRenameModal.tsx:23]: I'd really love to see this refactored into a logic... plus tests, etc.

**Review by mariusandra — COMMENTED** [review]
Retroactive review. All seems fine, though I'd really love to see the modal shown/hidden info to move into a logic as well. `entityFilterLogic` perhaps. I keep repeating myself, but if even for little local states like this, we store them in a logic, we get great benefits. Such as not having to rewrite this later when asked, having all the logic in an app follow the same pattern, easier testing, no prop passing, etc...

**Comment by alexkim205** [comment]
Yup I can move this into a logic and add some tests as well. Tracking here #6153

**Comment by charlescook-ph** [comment]
@alexkim205 Let me know if it's more appropriate to open a new issue for this, but I'm having some trouble with viewing the renamed filters in a few cases. [This is the graph I'm using](https://app.posthog.com/insights?insight=TRENDS&events=%5B%5D&actions=%5B%7B%22id%22%3A%2211036%22%2C%22math%22%3Anull%2C%22name%22%3A%22Marketing%3A%20Cloud%20signup%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A0%2C%22properties%22%3A%5B%7B%22key%22%3A%22utm_source%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22google%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22custom_name%22%3A%22Google%20(cloud)%22%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%2211037%22%2C%22math%22%3Anull%2C%22name%22%3A%22Marketing%3A%20Self%20host%20(free)%20signup%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A1%2C%22properties%22%3A%5B%7B%22key%22%3A%22utm_source%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22google%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22custom_name%22%3A%22Google%20(self%20host%20free)%22%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%2211036%22%2C%22math%22%3Anull%2C%22name%22%3A%22Marketing%3A%20Cloud%20signup%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A2%2C%22properties%22%3A%5B%7B%22key%22%3A%22utm_source%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22twitter%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22custom_name%22%3A%22Twitter%20(cloud)%22%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%2211037%22%2C%22math%22%3Anull%2C%22name%22%3A%22Marketing%3A%20Self%20host%20(free)%20signup%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A3%2C%22properties%22%3A%5B%7B%22key%22%3A%22utm_source%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22twitter%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22custom_name%22%3A%22Twitter%20(self%20host%20free)%22%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%2211036%22%2C%22math%22%3Anull%2C%22name%22%3A%22Marketing%3A%20Cloud%20signup%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A4%2C%22properties%22%3A%5B%7B%22key%22%3A%22utm_source%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22linkedin%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22custom_name%22%3A%22LinkedIn%20(cloud)%22%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%2211037%22%2C%22math%22%3Anull%2C%22name%22%3A%22Marketing%3A%20Self%20host%20(free)%20signup%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A5%2C%22properties%22%3A%5B%7B%22key%22%3A%22utm_source%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22linkedin%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22custom_name%22%3A%22LinkedIn%20(self%20host%20free)%22%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%2211038%22%2C%22math%22%3Anull%2C%22name%22%3A%22Marketing%20Self%20host%20(paid)%20conversion%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A6%2C%22properties%22%3A%5B%7B%22key%22%3A%22utm_source%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22google%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22custom_name%22%3A%22Google%20(self%20host%20paid)%22%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%2211038%22%2C%22math%22%3Anull%2C%22name%22%3A%22Marketing%20Self%20host%20(paid)%20conversion%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A7%2C%22properties%22%3A%5B%7B%22key%22%3A%22utm_source%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22twitter%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22custom_name%22%3A%22Twitter%20(self%20host%20paid)%22%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%2211038%22%2C%22math%22%3Anull%2C%22name%22%3A%22Marketing%20Self%20host%20(paid)%20conversion%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A8%2C%22properties%22%3A%5B%7B%22key%22%3A%22utm_source%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22linkedin%22%2C%22operator%22%3A%22icontains%22%7D%5D%2C%22custom_name%22%3A%22LinkedIn%20(self%20host%20paid)%22%2C%22math_property%22%3Anull%7D%5D&display=ActionsLineGraph&interval=week&date_from=-30d&new_entity=%5B%5D&properties=%5B%5D&filter_test_accounts=true#fromItem=403681&fromItemName=Paid%20Ads%20Conversions%20by%20UTM%20source&fromDashboard=7040) as an example. 

--------

**Details Table under a graph**

The names are so truncated that I can't view them unless I make them very short names (and I can't change the column widths on the table to accommodate. 

Screenshot 2021-10-13 at 10 01 37

**Hovering over the graph series - logged in**

Again when I hover over the graph series, the names are really truncated so I don't really know what each one refers to.

Screenshot 2021-10-13 at 09 55 04

**Hovering over the graph series - shared dashboard**

In a shared dashboard, it looks like the names pulled through are just of the queries and not the actual graph series (where I have queries with filters applied)? Not sure if this is intended behavior, but intuitively I would have expected it to behave in the same way as when viewing the graph in my account. Again the truncation makes it hard to make sense of the names. 

Screenshot 2021-10-13 at 09 56 11

**Pull request merged**